### PR TITLE
Fix some dynamic routes handling

### LINKED ISSUE #4191: Custom backend route prefix not used in robots.txt
### Affected version(s)

4.13

### Description

If you set your contao to have a custom backend route prefix, as described here https://docs.contao.org/manual/en/system/settings/#additional-back-end-settings

The disallow entry in the robots.txt is not adjusted correctly to the new route. As you can see here `contao` is hardcoded in there
https://github.com/contao/contao/blob/c904c5624e7f813980942cbf9dc5386405c0ee27/core-bundle/src/EventListener/RobotsTxtListener.php#L59

### PR DESCRIPTION
**Pull request by aschempp**
Fixes https://github.com/contao/contao/issues/4191

Also found a few other issues by chance. Instead of injecting the router into the `ContaoFramework`, we could also throw an appropriate exception and handle it in the `PrettyErrorScreenHandler`. But that would disable the redirect `dev` mode, but we could also render a template to explain instead of a plain redirect.

### CODE REVIEW

**Review by leofeyer — DISMISSED** [review]

**Comment by leofeyer** [comment]
You also have to adjust the tests, though. 🙈

File: core-bundle/src/Resources/config/services.yml [hunk 1]
```diff
@@ -1119,7 +1120,12 @@ services:
             version: 4.13
             message: Using the "%alias_id%" service ID has been deprecated and will no longer work in Contao 5.0. Please use "contao.csrf.token_manager" instead.
 
-    Contao\CoreBundle\Framework\ContaoFrameworkInterface: '@contao.framework'
+    Contao\CoreBundle\Framework\ContaoFrameworkInterface:
+        alias: contao.framework
+        deprecated:
+            package: contao/core-bundle
+            version: 4.7
+            message: Using the "%alias_id%" service ID has been deprecated and will no longer work in Contao 5.0. Please use "contao.framework" instead.
```
Comment (fritzmg) [inline core-bundle/src/Resources/config/services.yml]: This will need to be backported to Contao 4.9 as well.

Comment (aschempp) [reply core-bundle/src/Resources/config/services.yml]: see https://github.com/contao/contao/pull/4199

**Comment by aschempp** [comment]
no idea why tests fail, maybe @m-vo can tell?

**Comment by m-vo** [comment]
> no idea why tests fail, maybe @m-vo can tell?

The error happens when the code is accessing `imageSize` via the legacy `File` class. If you add the following to the test, the error disappears:

```php
// FigureRendererTest#testRendersContaoTemplate()`
// …
        $imageFactory = $this->createMock(ImageFactoryInterface::class);
        $imageFactory
            ->method('create')
            ->willReturn($this->createMock(ImageInterface::class))
        ;

        $container->set('contao.image.factory', $imageFactory);
```

This side effect is likely unrelated to this PR but only appears with it.

**Comment by m-vo** [comment]
https://github.com/aschempp/contao/pull/8

File: core-bundle/src/Framework/ContaoFramework.php [hunk 2]
```diff
@@ -358,7 +361,7 @@ private function validateInstallation(): void
         }
 
         if (!$this->getAdapter(Config::class)->isComplete()) {
-            throw new RedirectResponseException('/contao/install');
+            throw new RedirectResponseException($this->urlGenerator->generate('contao_install', [], UrlGeneratorInterface::ABSOLUTE_URL));
```
Comment (leofeyer) [inline core-bundle/src/Framework/ContaoFramework.php:364]: Why not inject the route prefix like in the `RobotsTxtListener` and redirect to `$this->routePrefix.'/install'`?

Comment (Toflar) [reply core-bundle/src/Framework/ContaoFramework.php:364]: What, why? The router is responsible to generate URLs?

Comment (leofeyer) [reply core-bundle/src/Framework/ContaoFramework.php:364]: We did not use the router before, either.

Comment (fritzmg) [reply core-bundle/src/Framework/ContaoFramework.php:364]: > We did not use the router before, either.

Which is is why this issue occurs in the first place.

**Review by leofeyer — APPROVED** [review]

**Pull request merged**